0001-Add-support-for-ESRT-loading-under-Xen.patch

```diff
@@ -0,0 +1,130 @@
+ include/linux/efi.h         | 10 ++++++++++
+ 3 files changed, 64 insertions(+), 9 deletions(-)
+
+diff --git a/drivers/firmware/efi/esrt.c b/drivers/firmware/efi/esrt.c
+index 2a2f52b017e736dd995c69e8aeb5fbd7761732e5..08722c7f4a2b0e98af661dce71e3fb690a770a98 100644
+--- a/drivers/firmware/efi/esrt.c
++++ b/drivers/firmware/efi/esrt.c
+@@ -247,19 +247,34 @@ void __init efi_esrt_init(void)
+ 	int rc;
+ 	phys_addr_t end;
+ 
+-	if (!efi_enabled(EFI_MEMMAP))
+-		return;
+-
+ 	pr_debug("esrt-init: loading.\n");
+ 	if (!esrt_table_exists())
+ 		return;
+ 
+-	rc = efi_mem_desc_lookup(efi.esrt, &md);
+-	if (rc < 0 ||
+-	    (!(md.attribute & EFI_MEMORY_RUNTIME) &&
+-	     md.type != EFI_BOOT_SERVICES_DATA &&
+-	     md.type != EFI_RUNTIME_SERVICES_DATA)) {
+-		pr_warn("ESRT header is not in the memory map.\n");
++	if (efi_enabled(EFI_MEMMAP)) {
++		if (efi_mem_desc_lookup(efi.esrt, &md) < 0 ||
++		    (!(md.attribute & EFI_MEMORY_RUNTIME) &&
++		     md.type != EFI_BOOT_SERVICES_DATA &&
++		     md.type != EFI_RUNTIME_SERVICES_DATA)) {
++			pr_warn("ESRT header is not in the memory map.\n");
++			return;
++		}
++	} else if (IS_ENABLED(CONFIG_XEN_EFI) && efi_enabled(EFI_PARAVIRT)) {
++		if (!xen_efi_mem_desc_lookup(efi.esrt, &md)) {
++			pr_warn("Failed to lookup ESRT header in Xen memory map\n");
++			return;
++		}
++
++		/* Recent Xen versions relocate the ESRT to memory of type
++		 * EfiRuntimeServicesData, which Xen will not reuse.  If the ESRT
++		 * is not in EfiRuntimeServicesData memory, it has not been reserved
++		 * by Xen and might be allocated to other guests, so it cannot
++		 * safely be used. */
++		if (md.type != EFI_RUNTIME_SERVICES_DATA) {
++			pr_warn("Xen did not reserve ESRT, ignoring it\n");
++			return;
++		}
++	} else {
+ 		return;
+ 	}
+ 
+diff --git a/drivers/xen/efi.c b/drivers/xen/efi.c
+index d1ff2186ebb48a7c0981ecb6d4afcbbb25ffcea0..f89d6f9464d83823b13960851c031b944bf5e1b6 100644
+--- a/drivers/xen/efi.c
++++ b/drivers/xen/efi.c
+@@ -26,6 +26,7 @@
+ 
+ #include <xen/interface/xen.h>
+ #include <xen/interface/platform.h>
++#include <xen/page.h>
+ #include <xen/xen.h>
+ #include <xen/xen-ops.h>
+ 
+@@ -40,6 +41,35 @@
+ 
+ #define efi_data(op)	(op.u.efi_runtime_call)
+ 
++bool xen_efi_mem_desc_lookup(u64 phys_addr, efi_memory_desc_t *out_md)
++{
++	struct xen_platform_op op = {
++		.cmd = XENPF_firmware_info,
++		.u.firmware_info = {
++			.type = XEN_FW_EFI_INFO,
++			.index = XEN_FW_EFI_MEM_INFO,
++			.u.efi_info.mem.addr = efi.esrt,
++		}
++	};
++	efi_memory_desc_t md;
++	union xenpf_efi_info *info = &op.u.firmware_info.u.efi_info;
++	int rc;
++
++	rc = HYPERVISOR_platform_op(&op);
++	if (rc) {
++		pr_warn("Could not obtain information on address %llu from Xen: "
++			"error %d\n", phys_addr, rc);
++		return false;
++	}
++
++	memset(&md, 0, sizeof(md));
++	md.attribute = info->mem.attr;
++	md.type = info->mem.type;
++	md.num_pages = info->mem.size >> XEN_PAGE_SIZE;
++	md.phys_addr = info->mem.addr;
++	return true;
```

Comment:
memcpy(md_out, md, sizeof(md)); ?

Reply:
Oops… that would explain why it doesn’t work :rofl:

Now fixed.